primitives/src/v5/mod.rs

```diff
@@ -1234,22 +1261,25 @@ pub enum DisputeStatement {
 impl DisputeStatement {
 	/// Get the payload data for this type of dispute statement.
 	pub fn payload_data(&self, candidate_hash: CandidateHash, session: SessionIndex) -> Vec<u8> {
-		match *self {
+		match self {
 			DisputeStatement::Valid(ValidDisputeStatementKind::Explicit) =>
 				ExplicitDisputeStatement { valid: true, candidate_hash, session }.signing_payload(),
 			DisputeStatement::Valid(ValidDisputeStatementKind::BackingSeconded(
 				inclusion_parent,
 			)) => CompactStatement::Seconded(candidate_hash).signing_payload(&SigningContext {
 				session_index: session,
-				parent_hash: inclusion_parent,
+				parent_hash: *inclusion_parent,
 			}),
 			DisputeStatement::Valid(ValidDisputeStatementKind::BackingValid(inclusion_parent)) =>
 				CompactStatement::Valid(candidate_hash).signing_payload(&SigningContext {
 					session_index: session,
-					parent_hash: inclusion_parent,
+					parent_hash: *inclusion_parent,
 				}),
 			DisputeStatement::Valid(ValidDisputeStatementKind::ApprovalChecking) =>
 				ApprovalVote(candidate_hash).signing_payload(session),
+			DisputeStatement::Valid(ValidDisputeStatementKind::ApprovalCheckingV2(
+				candidate_hashes,
+			)) => ApprovalVoteMultipleCandidates(candidate_hashes.clone()).signing_payload(session),
```

Comment:
This should ensure that `candidate_hash` is included in `candidate_hashes`. 

We could also save a bit of space by having the `ValidDisputeStatementKind::ApprovalCheckingV2` include only the _other_ candidate hashes and the index in the vec where the candidate hash in question should be included. Maybe overcomplicated, but `payload_data` returns a `Vec<u8>` unconditionally which is a problem. i.e. we can't represent a failure path with the current interface.

As it stands, we'd allow any set of candidate hashes to be treated as a valid dispute vote for any candidate.

Pseudocode
```rust
ValidDisputeStatementKind::ApprovalCheckingV2(candidate_hashes, index) => {
    let mut final_candidates = candidate_hashes.clone();
    let index = min(candidate_hashes.len(), index);
    final_candidates.insert(index, candidate_hash);
    signing_payload(final_candidates, session);
}
```

Reply:
Agree, this interface wasn't yet cleaned, it was just in a state that worked on the happy-path, I was thinking about making candidate_hash  an `Option<CandidateHash>` and when is Some we just added to the payload.

What do you think about that ?

Reply:
Opted instead for throwing an error  if the provided candidate_hash is not in the vec of candidate_hashes, I think that keeps the API easy to use and prevents accidental misusage.  

See here: https://github.com/paritytech/polkadot/pull/7554/commits/2271c7148b8c2dcc28833df0fb797461cab96c16#diff-e876407804ef19c762a4fe91db962a8eea5185a3aee42ed76b60fb8b549beb03R1288, let me know what you think.

Reply:
It seems fine, so long as we throw an error in dispute handling as normally done.

I do feel this part of the code could be much better optimized.